react/features/audio-mode/middleware.js

```diff
@@ -0,0 +1,56 @@
+import { AudioMode } from '../base/react-native';
+
+import { APP_WILL_MOUNT } from '../app';
+import {
+    CONFERENCE_FAILED,
+    CONFERENCE_LEFT,
+    CONFERENCE_WILL_JOIN
+} from '../base/conference';
+
+import { MiddlewareRegistry } from '../base/redux';
+
+/**
+ * Middleware that captures conference actions and sets the correct audio
+ * mode based on the type of conference.  Audio-only conferences don't
+ * use the speaker by default, and video conferences do.
+ *
+ * @param {Store} store - Redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register(store => next => action => {
+    switch (action.type) {
+    case APP_WILL_MOUNT: {
+        AudioMode.setMode(AudioMode.DEFAULT)
+            .catch(err => {
+                console.warn(`Error setting audio mode: ${err}`);
+            });
+        break;
+    }
+    case CONFERENCE_WILL_JOIN: {
+        let mode;
+        const state = store.getState()['features/base/conference'];
+
+        if (state.audioOnly) {
+            // TODO(saghul): Implement audio-only mode
+            mode = AudioMode.AUDIO_CALL;
+        } else {
+            mode = AudioMode.VIDEO_CALL;
+        }
+
+        AudioMode.setMode(mode)
+            .catch(err => {
+                console.warn(`Error setting audio mode: ${err}`);
+            });
+        break;
+    }
+    case CONFERENCE_FAILED:
+    case CONFERENCE_LEFT:
+        AudioMode.setMode(AudioMode.DEFAULT)
+            .catch(err => {
+                console.warn(`Error setting audio mode: ${err}`);
+            });
+        break;
```

Comment:
1. The branch CONFERENCE_FAILED/LEFT is the same as APP_WILL_MOUNT. For simplicity's sake, I prefer them  expressed as a single branch. Otherwise, I see they are different branches and I start desperately looking for a difference. At the end the difference is just the `{}` around the ALL_WILL_MOUNT branch.

2. All branches are basically, choose an AudioMode constant, invoke AudioMode.setMode with the chosen constant, catch an error, print an error. I prefer that logic to be expresses with as less branching as possible and as much common logic implemented in a common place as possible. So the branching is really just in "choose an AudioMode constant" and the remaining logic is the same.